---
title: "Get RVIS"
author: Kristoforus 
date: 29/07/2019
output: html_notebook
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggplot2)
library(dplyr)
library(MASS)

setwd(dirname(rstudioapi::getActiveDocumentContext()$path)) # set working directory as directory of current file
```

# Summary
This file transforms the raw ClinVar VCF into a CSV containing domain-level pathogenic and benign mutation counts.  

Input: 
- CSV of the domain-specific coordinate of every unique mutations (downstream_analysis/exac_data/all_mutations_domain_specific.csv)
Output: 
- Domain level RVIS scores (downstream_analysis/RVIS_analysis/RVIS_domain.csv) 
- Protein level RVIS scores (downstream_analysis/RVIS_analysis/RVIS_prot.csv)

NOTE: The number of variants commented on the chunks (e.g.  line 31, 33) is obtained from a trial analysis involving approx 6,900 proteins conducted on 29/07/2019. 


# Load files
```{r}
df <- read.csv("../exac_data/all_mutations_domain_specific.csv") # 2,862,793
df$AF <- as.character(df$AF) %>% as.numeric()
df[is.na(df$AF),]$AF <- 0. #1,493,667 rows with no mutations 

result.df <- df[,c(1,3)] %>% unique() 
result.df <- result.df[order(result.df$ENSG, result.df$cluster_membership),]
```

# Reduce the ExAC data with all variants
The ExAC data frame of all unique protein coding mutation as rows (above) is reduced to a data frame of number of mutations for every domain. 

This is ran in parallel on server using
- downstream_analysis/RVIS_analysis/reduce_final_all_variants.R
with input 
- downstream_analysis/exac_data/all_mutations_domain_specific.csv
to obtain 
- downstream_analysis/RVIS_analysis/all_mutations_domain_specific_reduced.csv

# Load reduced data frame
```{r}
reduced.df <- read.csv("all_mutations_domain_specific_reduced.csv")
reduced.df <- reduced.df[order(reduced.df$ENSG, reduced.df$cluster_membership),] # 36,469
reduced.df <- reduced.df[reduced.df$cluster_size > 20,] # all domains with less that 20 residues in it (32,312)
```

# Filter for common variants from complete data frame of all variants
```{r}
# We consider mutations with MAF > 0.01% as common mutations
df.comvar <- df[df$AF  > 0.0001,] # 146,757 common mutations 
df.comvar <- df.comvar[order(df.comvar$ENSG, df.comvar$cluster_membership, -df.comvar$AF),]
df.comvar$temp <- paste(df.comvar$ENSG, df.comvar$residue_position, sep = "-")
df.comvar.MAF <- df.comvar[!duplicated(df.comvar$temp),] # 120,632 unique residue position harbouring (at least one) mutation 
func <- unique(df.comvar.MAF$variant)[c(1,3,5,7,12)] 
df.comvar.MAF.func <- df.comvar.MAF[df.comvar.MAF$variant %in% func,] # 116,689 unique residues harbouring (at least one) functional mutation  
```

# How many proteins and domains harbours 0 mutation
```{r}
df.comvar.MAF.func$ENSG %>% unique %>% length # 6,923 proteins have functional mutations with MAF > 0.0001
reduced.df$ENSG %>% unique %>% length # 6,983 proteins have mutations (all kind)

df.comvar.MAF.func[c(1,3)] %>% unique %>% nrow # 29,030 protein domains have functional mutations with MAF > 0.0001
reduced.df %>% nrow() # 32,312 protein domains have mutations (all kind)
```

# Distributions of number of mutations in each protein and each domain 
```{r}
print('functional mutation counts for every protein and domain')
count(df.comvar.MAF.func, ENSG)$n %>% summary
count(df.comvar.MAF.func, ENSG, cluster_membership)$n %>% summary
print('all mutation counts for every protein and domain')
aggregated.ENSG <- reduced.df %>% group_by(ENSG) %>% summarise(n.mut = sum(n.mut))  
aggregated.ENSG$n.mut %>% summary
reduced.df$n.mut %>% summary
```

# Merging the functional mutation counts and all mutation counts together
```{r} 
reduced.df # 33,663 number of unique ENSG-cluster pairs 
reduced.df.comvar = count(df.comvar.MAF.func, ENSG, cluster_membership) # 29,030 ENSG-cluster pairs with mutations
colnames(reduced.df.comvar) <- c("ENSG", "cluster_membership", "com_var_count")
reduced.combined = merge(reduced.df, reduced.df.comvar, by=c('ENSG','cluster_membership'), all = TRUE)
reduced.combined[is.na(reduced.combined$com_var_count),]$com_var_count <- 0
reduced.combined <- reduced.combined[!is.na(reduced.combined$n.mut),] # 32,312 
```


# Regress number of all mutation with number of common mutations for domains and calculate studentised residuals
```{r}
ggplot(reduced.combined, aes(x=n.mut, y=com_var_count)) + geom_point()
domain.lm <- lm(n.mut ~ com_var_count, reduced.combined)
summary(domain.lm)
reduced.combined$RVIS.domain <- studres(domain.lm)
reduced.combined
```


# Regress number of all mutation with number of common mutations for genes and calculate studentised residuals
```{r}
reduced.combined.prot <- reduced.combined %>% group_by(ENSG) %>% summarise(n.mut = sum(n.mut), com_var_count = sum(com_var_count))  
ggplot(reduced.combined.prot, aes(x=n.mut, y=com_var_count)) + geom_point()
prot.lm <- lm(n.mut ~ com_var_count, reduced.combined.prot)
summary(prot.lm)
reduced.combined.prot$RVIS.prot <- studres(prot.lm)
reduced.combined.prot
```

# Distribution of RVIS values
```{r}
ggplot(reduced.combined, aes(RVIS.domain)) + geom_histogram()
reduced.combined$RVIS.domain %>% summary
ggplot(reduced.combined.prot, aes(RVIS.prot)) + geom_histogram()
reduced.combined.prot$RVIS.prot %>% summary
```

# WRITE RVIS results
```{r}
write.csv(reduced.combined, "RVIS_domain.csv", row.names = FALSE)
write.csv(reduced.combined.prot, "RVIS_prot.csv", row.names = FALSE)
```





































